compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -328,8 +333,13 @@ trait ConstraintHandling {
         || {
           constraint = c1
           val TypeBounds(lo, hi) = constraint.entry(param): @unchecked
-          isSub(lo, hi)
+          isSub(lo, hi) && {
+            // isSub may have introduced new constraints
+            replaceIfEqualBounds
```

Comment:
That's an expensive test. Consider replacing it with 
```scala
  def replaceIfEqualBounds(lo: Type, hi: Type): Boolean = // same as before
 ...
  isSubLoHi && {
    replaceIfEqualBounds(lo, hi)
    true
  }
```

Reply:
Ah it turns out that doesn't work since we would check whether the bounds we had prior to doing `isSub` are equal, hence missing potential updates in the constraints.

We can do this though
```scala
constraint = c1
val TypeBounds(lo, hi) = constraint.entry(param): @unchecked
isSub(lo, hi) && {
  if constraint ne c1 then // isSub introduced new constraints
    val TypeBounds(lo, hi) = constraint.nonParamBounds(param)
    replaceIfEqualBounds(lo, hi)
  true
}
```

Reply:
I'm investigating whether `replace` could do this kind of transitive replacing itself, instead of doing the check here.